File: algorithm/matrix/pascaltriangle/scala/SolutionII.scala
package pascaltriangles

import java.util

/**
  * Created by davidtan on 11/5/16
  *https://leetcode.com/problems/pascals-triangle-ii/
  * * Given an index k, return the kth row of the Pascal's triangle.

For example, given k = 3,
Return [1,3,3,1].

Note:
Could you optimize your algorithm to use only O(k) extra space?
  */
object SolutionII {

  def getRow(rowIndex: Int): util.ArrayList[Int] = {

    val result = new util.ArrayList[Int]()

    if (rowIndex < 0) return result

    result.add(1) // col = 0

    for (row <- 1 to rowIndex) {

      for (col <- result.size - 2 to 0 by -1) {

        result.set(col + 1, result.get(col) + result.get(col + 1))
      }

      result.add(1) // col is last
    }

    result
  }

  def main(args: Array[String]) = {
    println(getRow(3).toArray.toList) //List(1, 3, 3, 1)
  }


}

File: algorithm/matrix/pascaltriangle/java/SolutionII.java
package pascaltriangleII;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by davidtan on 11/4/16.
 * 119. Pascal's Triangle II
 * <p>
 * Given an index k, return the kth row of the Pascal's triangle.
 * <p>
 * For example, given k = 3,
 * Return [1,3,3,1].
 * <p>
 * Note:
 * Could you optimize your algorithm to use only O(k) extra space?
 */
public class SolutionII {

    public List<Integer> getRow(int rowIndex) {

        ArrayList<Integer> result = new ArrayList<>();

        if (rowIndex < 0) return result;

        result.add(1);

        for (int row = 1; row <= rowIndex; row++) {

            for (int col = result.size() - 2; col >= 0; col--) {

                result.set(col + 1, result.get(col) + result.get(col + 1));

            }

            result.add(1);

        }

        return result;

    }
    public static void main(String[] args){
        SolutionII sol = new SolutionII();
        System.out.println(sol.getRow(3));//[1, 3, 3, 1]
    }
}
